### PR TITLE
[#IOPID-268] Add FastLogin api for LC validation

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

Create  a new function with LC logic, that verify a lollipop request and returns a `SAMLResponse`.

#### Motivation and Context

To complete a fast-login request the Lollipop signature must be validated by a LC. This function is the LC implementation for this use case.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: api/internal.yaml.template [hunk 1]
```diff
@@ -25,12 +25,141 @@ paths:
           description: Not Found
         '500':
           description: Internal Server error
+  /api/v1/fast-login:
```
Comment (gquadrati) [inline api/internal.yaml.template]: `api/v1` is already defined in servers.url. 

Furthermore, servers.url is not correct, it should be `https://io-p-weu-fast-login-fn.azurewebsites.net` (as @Garma00 spotted out in hi PR https://github.com/pagopa/io-functions-fast-login/pull/8)

File: api/internal.yaml.template [hunk 2]
```diff
@@ -25,12 +25,141 @@ paths:
...
+          description: "Success response with SAMLResponse"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/FastLoginSAML"
 components:
+  securitySchemes:
+    ApiKeyAuth:
+      type: apiKey
+      description: This is an example of authentication. The LC can choose their custom auth method
```
Comment (gquadrati) [inline api/internal.yaml.template]: The description does not fit very well here

File: api/internal.yaml [hunk 3]
```diff
@@ -25,7 +25,83 @@ paths:
...
+          required: true
+          schema:
+            $ref: '#/components/schemas/LollipopSignature'
+      responses:
+        '200':
+          description: Success response with SAMLResponse
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/FastLoginSAML'
```
Comment (gquadrati) [inline api/internal.yaml]: It can also return `400` (`RequiredHeaderMiddleware` error), `401` and `500`

File: api/internal.yaml [hunk 4]
```diff
@@ -34,3 +110,60 @@ components:
...
+    AssertionRefSha512:
+      type: string
+      pattern: ^(sha512-[A-Za-z0-9-_=]{1,88})$
+    FastLoginSAML:
+      type: object
+      properties:
+        saml_response:
+          type: string
+      required:
+        - saml_response
```
Comment (gquadrati) [inline api/internal.yaml]:
```suggestion
    FastLoginSAML:
      type: object
      properties:
        saml_response:
          type: string
          minLength: 1
      required:
        - saml_response
```

File: src/middlewares/request.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+export const RequiredHeaderMiddleware: <T>(
+  schema: Decoder<unknown, T>,
+  headerName: NonEmptyString
+) => RTE.ReaderTaskEither<H.HttpRequest, H.HttpBadRequestError, T> = <T>(
+  schema: Decoder<unknown, T>,
+  headerName: NonEmptyString
+) =>
+  flow(
+    req => req.headers,
+    lookup(headerName),
```
Comment (gquadrati) [inline src/middlewares/request.ts:19]: ❤️

File: src/utils/lollipop/crypto.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import { getCustomVerifyWithEncoding } from "./httpSignature.verifiers";
+
+export const validateHttpSignatureWithEconding = (
+  dsaEncoding: crypto_lib.DSAEncoding
+) => (
+  request: H.HttpRequest,
+  assertionRef: AssertionRef,
+  publicKey: JwkPublicKey,
+  body?: string
+): TE.TaskEither<Error, true> =>
```
Comment (gquadrati) [inline src/utils/lollipop/crypto.ts]: We could move to object param

```suggestion
export const validateHttpSignatureWithEconding = (
  dsaEncoding: crypto_lib.DSAEncoding
) => ({
  request,
  assertionRef,
  publicKey,
  body
}: {
  readonly request: H.HttpRequest;
  readonly assertionRef: AssertionRef;
  readonly publicKey: JwkPublicKey;
  readonly body?: string;
}): TE.TaskEither<Error, true> =>
```

File: src/functions/fast-login.ts [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+          pipe(
+            validateHttpSignatureWithEconding("der")(...params),
+            TE.orElse(() =>
+              validateHttpSignatureWithEconding("ieee-p1363")(...params)
+            )
+          ),
+        TE.mapLeft(
+          () => new H.HttpUnauthorizedError("Invalid Lollipop Signature")
+        )
+      )
```
Comment (gquadrati) [inline src/functions/fast-login.ts]: Can we extract it into a function, for a better readability?

File: src/functions/fast-login.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+                  )
+                ],
+                RA.map(verifier => verifier(_.samlResponseDoc)),
+                TE.sequenceArray,
+                RTE.fromTaskEither
+              )
+            )
+          )
+        )
+      ),
```
Comment (gquadrati) [inline src/functions/fast-login.ts]: Can we make it a little more linear?

```suggestion
    RTE.bindTo("verifiedHeaders"),
    RTE.bindW("samlResponse", ({ verifiedHeaders }) =>
      RetrieveSAMLResponse(verifiedHeaders.lollipopHeaders)
    ),
    RTE.bindW("samlResponseDoc", ({ samlResponse }) =>
      pipe(
        TE.tryCatch(
          async () =>
            new DOMParser().parseFromString(
              samlResponse.saml_response,
              "text/xml"
            ),
          () => new H.HttpError("Error parsing the SAMLResponse")
        ),
        RTE.fromTaskEither
      )
    ),
    RTE.chainFirstW(({ verifiedHeaders, samlResponseDoc }) =>
      pipe(
        [
          getAssertionRefVsInRensponseToVerifier(
            verifiedHeaders.publicKey,
            verifiedHeaders.lollipopHeaders[ASSERTION_REF_HEADER_NAME]
          ),
          getAssertionUserIdVsCfVerifier(
            verifiedHeaders.lollipopHeaders["x-pagopa-lollipop-user-id"]
          )
        ],
        RA.map(verifier => verifier(samlResponseDoc)),
        TE.sequenceArray,
        RTE.fromTaskEither
      )
    ),
    RTE.map(({ samlResponse }) => samlResponse),
```

File: src/functions/fast-login.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+          ["x-pagopa-lollipop-auth"]: `Bearer ${lollipopHeaders["x-pagopa-lollipop-auth-jwt"]}` as LollipopAuthBearer
+        }),
+      () => new H.HttpError("Error calling the getAssertion endpoint")
+    ),
+    TE.chainW(
+      flow(
+        TE.fromEither,
+        TE.mapLeft(_ => new H.HttpError("Unexpected response from fn-lollipop"))
+      )
+    ),
```
Comment (gquadrati) [inline src/functions/fast-login.ts]:
```suggestion
    TE.chainEitherK(
      E.mapLeft(_ => new H.HttpError("Unexpected response from fn-lollipop"))
    ),
```

File: src/functions/__tests__/fast-login.test.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    expect(getAssertionMock).toBeCalled();
+    expect(E.isRight(result)).toBeTruthy();
+    if (E.isRight(result)) {
+      expect(result.right).toEqual(
+        expect.objectContaining({
+          body: { saml_response: aSAMLResponse },
+          statusCode: 200
+        })
+      );
+    }
```
Comment (gquadrati) [inline src/functions/__tests__/fast-login.test.ts]: We can simplify as

```suggestion
    expect(result).toEqual(
      E.right(
        expect.objectContaining({
          body: { saml_response: aSAMLResponse },
          statusCode: 200
        })
      )
    );
```

File: src/functions/__tests__/fast-login.test.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    })();
+    expect(getAssertionMock).not.toBeCalled();
+    expect(E.isLeft(result)).toBeTruthy();
+    if (E.isLeft(result)) {
+      expect(result.left).toEqual(
+        expect.objectContaining({
+          status: 400
+        })
+      );
+    }
```
Comment (gquadrati) [inline src/functions/__tests__/fast-login.test.ts]: ditto

```suggestion
 expect(result).toEqual(
      E.left(
        expect.objectContaining({
          status: 400
        })
      )
    );
```

File: package.json [hunk 12]
```diff
@@ -21,6 +21,7 @@
     "pregenerate": "npm-run-all api:bundle*",
     "generate": "npm-run-all generate:*",
     "generate:definitions-internal": "rimraf ./src/generated/definitions/internal && shx mkdir -p ./src/generated/definitions/internal && gen-api-models --api-spec ./api/internal.yaml --no-strict --out-dir ./src/generated/definitions/internal ",
+    "generate:definitions-fn-lollipop": "rimraf ./src/generated/definitions/fn-lollipop && shx mkdir -p ./src/generated/definitions/fn-lollipop && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-lollipop/master/openapi/external.yaml --no-strict --out-dir ./src/generated/definitions/fn-lollipop --request-types --response-decoders --client",
```
Comment (gquadrati) [inline package.json]: `generate:client-fn-lollipop` is more appropriate

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: src/functions/__tests__/fast-login.test.ts (hunk 11 above)
Comment (BurnedMarshal) [reply src/functions/__tests__/fast-login.test.ts]: fixed in [68eda23](https://github.com/pagopa/io-functions-fast-login/pull/9/commits/68eda2317ec8a2c8706c34c1cb666f5ea1f5a3ed)

File: src/functions/__tests__/fast-login.test.ts (hunk 10 above)
Comment (BurnedMarshal) [reply src/functions/__tests__/fast-login.test.ts]: Fixed in [68eda23](https://github.com/pagopa/io-functions-fast-login/pull/9/commits/68eda2317ec8a2c8706c34c1cb666f5ea1f5a3ed)

File: package.json (hunk 12 above)
Comment (BurnedMarshal) [reply package.json]: Fixed in 52e77cb

File: src/middlewares/request.ts (hunk 5 above)
Comment (BurnedMarshal) [reply src/middlewares/request.ts:19]: I've copy that from `io-sign`

File: src/functions/fast-login.ts (hunk 8 above)
Comment (BurnedMarshal) [reply src/functions/fast-login.ts]: Fixed in f37570c

File: src/functions/fast-login.ts (hunk 7 above)
Comment (BurnedMarshal) [reply src/functions/fast-login.ts]: Fixed in 28aacc1

File: src/utils/lollipop/crypto.ts (hunk 6 above)
Comment (BurnedMarshal) [reply src/utils/lollipop/crypto.ts]: Fixed in 28aacc1

File: api/internal.yaml.template (hunk 1 above)
Comment (BurnedMarshal) [reply api/internal.yaml.template]: Fixed in 3445b3b

File: api/internal.yaml.template (hunk 2 above)
Comment (BurnedMarshal) [reply api/internal.yaml.template]: Fixed in 3445b3b

File: api/internal.yaml (hunk 3 above)
Comment (BurnedMarshal) [reply api/internal.yaml]: Fixed in 3445b3b

**Review by gquadrati — APPROVED** [review]
LGTM

File: src/config.ts [hunk 13]
```diff
@@ -12,19 +12,37 @@ import { pipe } from "fp-ts/lib/function";
...
+import { withDefault } from "@pagopa/ts-commons/lib/types";
+import { NumberFromString } from "@pagopa/ts-commons/lib/numbers";
 
 // ----------------------------
 // Global app configuration
 // ----------------------------
+const GetAssertionConfig = t.type(
+  {
+    LOLLIPOP_GET_ASSERTION_API_KEY: NonEmptyString,
+    LOLLIPOP_GET_ASSERTION_BASE_URL: UrlFromString
```
Comment (gquadrati) [inline src/config.ts:25]: Names are different from the ones defined on [infra](https://github.com/pagopa/io-infra/blob/d664d473ffdd4cee8dfe482b2923786c8a9f4f69/src/domains/citizen-auth-app/07_function_fast_login.tf#L26)

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: src/config.ts (hunk 13 above)
Comment (BurnedMarshal) [reply src/config.ts:25]: Yes, IMHO should be changed in infra. `LC` prefix refers to the LollipopConsumer but in here is this function the LC and the parameters refers to the fn-lollipop.

Comment (BurnedMarshal) [reply src/config.ts:25]: https://github.com/pagopa/io-infra/pull/600

File: FastLogin/function.json [hunk 14]
```diff
@@ -0,0 +1,21 @@
+{
+  "bindings": [
+    {
+      "authLevel": "function",
+      "type": "httpTrigger",
+      "direction": "in",
+      "name": "req",
+      "route": "/api/v1/fast-login",
```
Comment (gquadrati) [inline FastLogin/function.json]:
```suggestion
      "route": "api/v1/fast-login",
```

File: src/config.ts [hunk 15]
```diff
@@ -12,19 +12,37 @@ import { pipe } from "fp-ts/lib/function";
...
+  t.interface({
+    APPINSIGHTS_INSTRUMENTATIONKEY: NonEmptyString,
+
+    COSMOS_CONNECTION_STRING: NonEmptyString,
+    COSMOS_DB_NAME: NonEmptyString,
 
-  COSMOS_CONNECTION_STRING: NonEmptyString,
-  COSMOS_DB_NAME: NonEmptyString,
+    // Default is 10 sec timeout
+    FETCH_TIMEOUT_MS: withDefault(NumberFromString, 10000),
```
Comment (gquadrati) [inline src/config.ts]: This is gonna fail if env var is not defined. We can refactor as:

```suggestion
    // Default is 10 sec timeout
    FETCH_TIMEOUT_MS: withDefault(t.string, "10000").pipe(NumberFromString),
```

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: FastLogin/function.json (hunk 14 above)
Comment (BurnedMarshal) [reply FastLogin/function.json]: fixed in 9c73cd6

File: src/config.ts (hunk 15 above)
Comment (BurnedMarshal) [reply src/config.ts]: Fixed in 9c73cd6

**Review by gquadrati — APPROVED** [review]

**Pull request merged**